[nrf52840] Improvements to USB-CDC-UART

darconeous opened the pull request:
This change addresses some issues that was noticed while working to bring up the OpenThread NCP on the nRF52840. The most important change is the introduction of the macro `USB_HOST_UART_CONFIG_DELAY_MS`, which will slightly delay the output of queued data once the fake UART is opened. On some slow Linux systems, the USB-CDC-UART driver sends the data so quickly after the port opens that the application hasn't had time to get the TTY properly configured using tcsetattr(). While this is technically a host-side problem, fixing this problem host-side is architecturally infeasible in many cases. Ensuring the reliable delivery of the first bytes of the HDLC stream from the NCP after a reset is critical to the host driver being able to identify that a reset has indeed occurred. Waiting a few milliseconds after the port is opened allows the reset reason frame to be reliably transmitted.

rlubos commented on examples/platforms/nrf52840/usb-cdc-uart.c:
```diff
@@ -363,25 +378,26 @@ otError otPlatUartDisable(void)
+    otEXPECT_ACTION(sUsbState.mTransferInProgress == false, error = OT_ERROR_BUSY);
+    otEXPECT_ACTION(sUsbState.mTxBuffer == NULL, error = OT_ERROR_BUSY);
+
     if (!isPortOpened())
```
Shouldn't `sUsbState.mOpenTimestamp` be checked here as well? If there were no queued data and we call `otPlatUartSend` just after the port is opened we will send it without waiting the `USB_HOST_UART_CONFIG_DELAY_MS `.

I can see though, that in case we check for this in `otPlatUartSend`, the time checking logic should be protected against timer overflow, something like:
```
bool hasPortOpenDelayPassed(void)
{
    int32_t timeDiff = otPlatAlarmMilliGetNow() - sUsbState.mOpenTimestamp;

    return (timeDiff < 0) || (timeDiff > USB_HOST_UART_CONFIG_DELAY_MS);
}
```

darconeous replied:
I wasn't too worried about that case because it is effectively guaranteed not to happen with the way the NCP code is currently written for the specific scenario we are trying to prevent (failing to receive the reset reason).

Adding the check would end up causing a periodic delay of 10ms every few weeks (due to roll-over), but that seems so insignificant I can't really consider it a concern.

I'll go ahead and address and test that today so that we can get this merged in ASAP.

darconeous commented:
Ok, I've addressed those concerns in the latest commit. I've moved the "open delay passed" check into the static function `isPortOpen()`, since in all places where `isPortOpen()` is called also needed the "open delay passed" check (I verified). I feel like the code is even cleaner now with these changes.

I've tested the changes and they appear to work great.

@jwhui : Given the previous 👍  from @rlubos, do you think we could get this merged today so that we can update <https://openthread.io/guides/ncp/firmware> to include this change?

jwhui commented:
@darconeous, sounds good.  I'll wait for Travis to pass then merge.
